keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1201 @@
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation. But downgrading from an update that widens validation can be understood 
+semantically as an update that tightens validating.
+
+For example if they had unreletated issues with the new Kubernetes update and 
+wanted to downgrade:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers fail to update some resources, since they fail validation.
+
+If the CRD Validation Ratcheting feature were enabled the following occurs:
+
+1. User updates to new Kubernetes version
+2. New CRs that were previously disallowed are allowed to be stored
+3. User is disappointed with K8s version and downgrades
+4. Controllers are allowed to update their resources, as long as all their changes are valid.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This KEP is scoped to CRDs, but if [KEP-3937](https://github.com/kubernetes/enhancements/pull/3938) is implemented then we should expect it to also affect native type validation.
+
+#### Change in API Behavior
+
+This KEP changes assumptions that calls to Update will exhaustively validate
+every field of the CR. After this change, authors of CRs will need to do this test 
+with a CREATE rather than an UPDATE.
+
+NOTE: Errors that would have normally been thrown are converted into warnings,
+so authors using UPDATE to validate their CR still get feedback.
+
+### Risks and Mitigations
+
+#### Detection of Breaking Schema Changes is Different
+
+Some users might have a process to detect breaking schema changes by applying
+their CRD and poking an existing resource. This KEP would make that process
+no longer reliable.
+
+##### Use CRD-Schema-Checker
+
+[CRD-Schema-Checker](https://github.com/openshift/crd-schema-checker) may be
+used to lint the CRD schema for breaking changes before applying it to a cluster.
+
+##### Use New Objects
+
+Alternatively, users may create new objects rather than updating old ones to 
+test for breaking schema changes. This feature has no effect on the creation 
+code path.
+
+
+#### Some CRD Validations Are Not Ratchetable
+
+Some usages of the OpenAPI schema are not possible to reliably ratchet without
+creating confusing exceptions, or rewriting completely the validation code used
+by Kubernetes. 
+
+##### Blacklisted Validations
+
+To mitigate this, validation failures in rules which have an
+ancestor of one of these types will not be ratcheted:
```

Comment:
The `schema_options.go` responsible for the blacklisted validation types still directly call `NewSchemaValidator` rather than the `Sub(Property/Index)Validator` hooks which provide the ratcheting functionality.

Reply:
updated